File: KeyAlert_objc/KeyAlert_objc/KeyAlert/KeyAlert.m
//
//  KeyAlert.m
//  KeyAlert_objc
//
//  Created by KEY on 2018. 6. 14..
//  Copyright © 2018년 KEY. All rights reserved.
//  commit test

#import "KeyAlert.h"

#define default_Max_MsgLabel_width          [UIScreen mainScreen].bounds.size.width - 60
#define default_Max_MsgLabel_height         30
#define default_MsgLabel_Horizontal_Space   20
#define default_MsgLabel_vertical_Space     5

@implementation KeyAlert
@synthesize alertView;

+ (KeyAlert *)shared {
    static KeyAlert *sharedManager = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{    //라이프 사이클간 일회 실행
        sharedManager = [[self alloc] init];
    });
    return sharedManager;
}

- (instancetype)init {
    if (self = [super init]) {
        alertView = [[KeyAlertView alloc] initWithNibName:@"KeyAlertView" bundle:nil];
    }
    return self;
}



+ (void)alertShowWithMessage:(nullable NSString *)message
                      target:(nonnull UIViewController *)target
                buttonTitles:(nonnull NSArray *)buttonTitles
                    callBack:(nullable CallBackBlock)callBack {
    
    [self shared].alertView.message       = message;
    [self shared].alertView.buttonTitles  = buttonTitles;
    [self shared].alertView.callBackBlock = callBack;
    
    [self shared].alertView.modalPresentationStyle = UIModalPresentationOverCurrentContext;
    [target presentViewController:[self shared].alertView animated:NO completion:nil];
    
}

+ (void)toastWithMessage:(nullable NSString *)message
               completed:(nullable ToastCompleted)completed {
    
    UIWindow *window = [[UIApplication sharedApplication] keyWindow];
    
    UILabel *lbMsg = [[UILabel alloc] init];
    
    lbMsg.lineBreakMode = NSLineBreakByWordWrapping;
    lbMsg.numberOfLines = 0;
    lbMsg.font = [UIFont systemFontOfSize:13];
    lbMsg.textColor = [UIColor whiteColor];
    
    CGSize maxSize = CGSizeMake(default_Max_MsgLabel_width, default_Max_MsgLabel_height);
    CGRect expectedLabelRect = [message boundingRectWithSize:maxSize options:NSStringDrawingUsesLineFragmentOrigin attributes:@{ NSFontAttributeName:lbMsg.font } context:nil];
    expectedLabelRect.origin.x = default_MsgLabel_Horizontal_Space;
    expectedLabelRect.origin.y = default_MsgLabel_vertical_Space;
    lbMsg.frame = expectedLabelRect;
    lbMsg.text = message;
    
    float lbMsg_Width = lbMsg.frame.size.width;
    
    float toastPosisionX = ([UIScreen mainScreen].bounds.size.width - (lbMsg_Width + default_MsgLabel_Horizontal_Space * 2))/2;
    
    float toastPosisionY = [UIScreen mainScreen].bounds.size.height - 80;
    
    
    
    UIView *toast = [[UIView alloc] initWithFrame:CGRectMake(toastPosisionX,
                                                             [UIScreen mainScreen].bounds.size.height,
                                                             lbMsg_Width + default_MsgLabel_Horizontal_Space * 2,
                                                             lbMsg.frame.size.height + default_MsgLabel_vertical_Space * 2)];

    toast.backgroundColor = [UIColor grayColor];
    toast.layer.cornerRadius = 10;
    
    [toast addSubview:lbMsg];
    [window addSubview:toast];
    
    [UIView animateWithDuration:0.3 animations:^{
        toast.frame = CGRectMake(toast.frame.origin.x,
                                 toastPosisionY,
                                 lbMsg_Width + 40,
                                 toast.frame.size.height);
        toast.alpha = 1;
    } completion:^(BOOL finished){
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            [UIView animateWithDuration:0.3 animations:^{
                toast.frame = CGRectMake(toast.frame.origin.x,
                                         [UIScreen mainScreen].bounds.size.height,
                                         lbMsg_Width + 40,
                                         toast.frame.size.height);
                toast.alpha = 0;
            } completion:^(BOOL finished){
                [toast removeFromSuperview];
                if(completed)completed();
            }];
        });
    }];
    
}

+ (void)moveCenterPosision_x:(float)x
                           y:(float)y {
    
    [[self shared].alertView moveCenterPosision_x:x y:y];
    
}


+ (void)setPopupSize_width:(float)width
                   tHeight:(float)tHeight
                mMinHeight:(float)mMinHeight
                mMaxHeight:(float)mMaxHeight
                   bHeight:(float)bHeight {
    
    [[self shared].alertView setPopupSize_width:width
                                        tHeight:tHeight
                                     mMinHeight:mMinHeight
                                     mMaxHeight:mMaxHeight
                                        bHeight:bHeight];
    
}

+ (void)setContentLabelSpase_top:(float)top
                            left:(float)left
                          bottom:(float)bottom
                           right:(float)right {
    
    [[self shared].alertView setContentLabelSpase_top:top
                                                 left:left
                                               bottom:bottom
                                                right:right];
    
}

+ (void)setButtonSpase_btnTobtn:(float)btnTobtn
                   contentTobtn:(float)contentTobtn {
    
    [[self shared].alertView setButtonSpase_btnTobtn:btnTobtn
                                        contentTobtn:contentTobtn];
    
}

+ (void)resetStruct {
    [[self shared].alertView resetStruct];
}
@end

File: KeyAlert_swift/KeyAlert_swift/KeyAlert/KeyAlert.swift
//
//  KeyAlert.swift
//  KeyAlert_swift
//
//  Created by KEY on 2018. 7. 6..
//  Copyright © 2018년 KEY. All rights reserved.
//

import UIKit

class KeyAlert {
    
    let default_Max_MsgLabel_width: CGFloat = UIScreen.main.bounds.size.width - 60
    let default_Max_MsgLabel_height: CGFloat = 30
    let default_MsgLabel_Horizontal_Space: CGFloat = 20
    let default_MsgLabel_vertical_Space: CGFloat = 5

    static let shared = KeyAlert()
    var alertView: KeyAlertView?
    
    init() {
        alertView = KeyAlertView(nibName: "KeyAlertView", bundle: nil)
    }
    
    func alertShow(message: String, target: UIViewController, buttonTitles:Array<String>, callBack: @escaping (Int, UIViewController) -> Swift.Void) -> Void {
        alertView?.message = message
        alertView?.buttonTitles = buttonTitles
        alertView?.CallBackBlock = callBack
        target.present(alertView!, animated: false, completion: nil)
    }
    
    func resetStruct() -> Void {
        alertView?.resetStruct()
    }
    
    func moveCenterPosision(x:CGFloat, y:CGFloat) -> Void {
        alertView?.moveCenterPosision(x: x, y: y)
    }
    
    func setPopupSize(width:CGFloat, tHeight:CGFloat, mMinHeight:CGFloat, mMaxHeight:CGFloat, bHeight:CGFloat) -> Void {
        alertView?.setPopupSize(width: width, tHeight: tHeight, mMinHeight: mMinHeight, mMaxHeight: mMaxHeight, bHeight: bHeight)
    }
    
    func setContentLabelSpase(top: CGFloat, left: CGFloat, bottom: CGFloat, right: CGFloat) -> Void {
        alertView?.setContentLabelSpase(top: top, left: left, bottom: bottom, right: right)
    }
    
    func setButtonSpase(btnTobtn: CGFloat, contentTobtn: CGFloat) -> Void {
        alertView?.setButtonSpase(btnTobtn: btnTobtn, contentTobtn: contentTobtn)
    }
    
    func toast(message: String, completed: @escaping () -> Swift.Void) -> Void {
        let window: UIWindow = UIApplication.shared.keyWindow!
        
        let lbMsg: UILabel = UILabel.init()
        
        lbMsg.lineBreakMode = .byWordWrapping
        lbMsg.numberOfLines = 0
        lbMsg.font = UIFont.systemFont(ofSize: 13)
        lbMsg.textColor = UIColor.white
        
        let maxSize: CGSize = CGSize(width: default_Max_MsgLabel_width,
                                     height: default_Max_MsgLabel_height)
        var expectedLabelRect = message.boundingRect(with: maxSize, options: .usesLineFragmentOrigin, attributes: [NSAttributedStringKey.font: lbMsg.font], context: nil)
        expectedLabelRect.origin.x = default_MsgLabel_Horizontal_Space
        expectedLabelRect.origin.y = default_MsgLabel_vertical_Space
        lbMsg.frame = expectedLabelRect
        lbMsg.text = message
        
        let lbMsg_Width = lbMsg.frame.size.width
        let toastPosisionX = (UIScreen.main.bounds.size.width - (lbMsg_Width + default_MsgLabel_Horizontal_Space * 2)) / 2
        let toastPosisionY = UIScreen.main.bounds.size.height - 80
        
        let toast: UIView = UIView(frame: CGRect(x: toastPosisionX,
                                                 y: UIScreen.main.bounds.size.height,
                                                 width: lbMsg_Width + default_MsgLabel_Horizontal_Space * 2,
                                                 height: lbMsg.frame.size.height + default_MsgLabel_vertical_Space * 2))
        toast.backgroundColor = UIColor.gray
        toast.layer.cornerRadius = 10
        
        toast.addSubview(lbMsg)
        window.addSubview(toast)
        
        UIView.animate(withDuration: 0.3, animations: {
            toast.frame = CGRect(x: toast.frame.origin.x,
                                 y: toastPosisionY,
                                 width: lbMsg_Width + 40,
                                 height: toast.frame.size.height)
            toast.alpha = 1
        }) { (finished) in
            DispatchQueue.main.asyncAfter(deadline: .now() + 2.0, execute: {
                UIView.animate(withDuration: 0.3, animations: {
                    toast.frame = CGRect(x: toast.frame.origin.x,
                                         y: UIScreen.main.bounds.size.height,
                                         width: lbMsg_Width + 40,
                                         height: toast.frame.size.height)
                }, completion: { (finished) in
                    toast.removeFromSuperview()
                    completed()
                })
            })
        }
        
        
    }
}
